### PR TITLE
fix for the issue #321

### PR DESCRIPTION
**Pull request by ipdemes**
To fix the issue when 2 stores with the shape (1,) (Futures) needs to be copied, I have created a `COPY` task. 
At this point `COPY` task is used only in the advanced indexing operation, however, it can replace `COPY` operation from the `unary_op`.  @manopapad @magnatelee what do you think?

FYI I plan to use this `COPY` task for another PR that implements logic for ND output regions for the `advanced_indexing` task. In this PR I will need to copy two fields with `pointN` type.

### CODE REVIEW

File: cunumeric/deferred.py [hunk 1]
```diff
@@ -624,13 +624,25 @@ def get_item(self, key):
...
+                if index_array.base.kind == Future:
+                    # TODO replace with .copy() call when copy() uses COPY task
+                    # instead of the UNARY_OP.COPY
+                    task = self.context.create_task(CuNumericOpCode.COPY)
+                    task.add_output(result.base)
+                    task.add_input(store)
+                    task.add_input(index_array.base)
+                    task.add_scalar_arg(True, bool)  # source indirect
+                    task.add_alignment(result.base, index_array.base)
+                    task.add_broadcast(store)
```
Comment (ipdemes) [inline cunumeric/deferred.py:636]: I don't like very much the idea of broadcasting input array, but I am not sure how to partition it when I have an indirection

**Comment by magnatelee** [comment]
@ipdemes I'm fine with extracting out the copy task from the unary op, as it deals with some special types other unary ops don't handle. But like I said in the other comment, you need to make sure that you never materialize any of the arguments one only one of the memories when you're calling the copy task.

File: cunumeric/deferred.py (hunk 1 above)
Comment (ipdemes) [reply cunumeric/deferred.py:636]: @magnatelee In this particular example I use `copy` task only when indirect field is a `Future`.  So the question is: how would I convert `Future` to the `Store`?

**Comment by ipdemes** [comment]
@magnatelee 
What do you mean by "you never materialize any of the arguments one only one of the memories when you're calling the copy task." ?
In this PR I am covering a particular use case of COPY task when indirect field is a `Future` and not trying to use this new `COPY` task for the `copy()` operation. I was planning create another PR where I change the `copy()` operation to use COPY task.  There I would try to use `Legate copy` when possible and use OPY task for some specific cases.

File: cunumeric/deferred.py (hunk 1 above)
Comment (magnatelee) [reply cunumeric/deferred.py:636]: It's done in two steps. First, you need to create a store of volume 1 that is backed by a region field. You could do it by passing `optimize_scalar=False` when creating the store: https://github.com/nv-legate/cunumeric/blob/branch-22.05/cunumeric/runtime.py#L415-L419 Then, you copy a future-backed store to that new store. I guess you might want to add a runtime method for doing it, as it's a common utility.

**Comment by magnatelee** [comment]
> @magnatelee
What do you mean by "you never materialize any of the arguments one only one of the memories when you're calling the copy task." ?
In this PR I am covering a particular use case of COPY task when indirect field is a Future and not trying to use this new COPY task for the copy() operation. I was planning create another PR where I change the copy() operation to use COPY task. There I would try to use Legate copy when possible and use OPY task for some specific cases.

If I understood your code correctly, it's possible that either the input or the output could be a regular store even when the indirection field was a future-backed one, right? For such cases, you shouldn't materialize that regular store on one of the memories, either with a manual broadcasting constraint or by launching the task with a future output (which will always be serialized). Like I suggested, you should instead covert those future-backed stores to regular ones and use the normal Legate copy. (I think the current copy implementation would still try to use a individual copy and not an index copy for such cases, but I can fix that, whereas if you launched the task with a broadcasting constraint, there's nothing I can do.)

File: cunumeric/deferred.py (hunk 1 above)
Comment (ipdemes) [reply cunumeric/deferred.py:636]: I see. How do I decide that store is a big store? Should I just always convert `Future` to `Store` and then use CopyLauncher in this case?

Comment (ipdemes) [reply cunumeric/deferred.py:636]: If I do that I wouldn't need to have a separate COPY task at all

Comment (magnatelee) [reply cunumeric/deferred.py:636]: Right. if any of the three stores can be a regular store, I think you should be converting the others to regular stores and using the Legate copy. (I thought the case this PR is addressing is when all three stores are backed by futures.)

Comment (ipdemes) [reply cunumeric/deferred.py:636]: Then I will close this PR and open another one that takes proposed approach.  Thank you!

**Comment by ipdemes** [comment]
Closing this PR with the plan to open another one for this issue

**Pull request closed without merging**